backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/incremental/utils/Geometry.java

```diff
@@ -114,10 +131,126 @@ private Position getIntersection(Position lineA, Position lineB, Position segmen
         return p;
     }
 
-    public double computeDistance(Position source, Position target) {
+    public static double computeDistance(Position source, Position target) {
         double ac = Math.abs(target.getY() - source.getY());
         double cb = Math.abs(target.getX() - source.getX());
         return Math.hypot(ac, cb);
     }
 
+    /**
+     * Returns the distance between the point p and the segment made of p1 and p2.
+     *
+     * @param p
+     *            the point
+     * @param p1
+     *            the first point of the segment
+     * @param p2
+     *            the second point of the segment
+     */
+    public static Position closestPointOnSegment(Position p, Position p1, Position p2) {
+        double p1pVectordx = p.getX() - p1.getX();
+        double p1pVectordy = p.getY() - p1.getY();
+        double p2p1Vectordx = p2.getX() - p1.getX();
+        double p2p1Vectordy = p2.getY() - p1.getY();
+
+        double dot = p1pVectordx * p2p1Vectordx + p1pVectordy * p2p1Vectordy;
+        double lengthSquare = p2p1Vectordx * p2p1Vectordx + p2p1Vectordy * p2p1Vectordy;
+        // param is the projection distance from p1 on the segment. This is the fraction of the segment that p is
+        // closest to.
+        double param = -1;
+        if (lengthSquare != 0) {
+            // if the line has 0 length
+            param = dot / lengthSquare;
+        }
+
+        Position nearestPoint = null;
+
+        if (param < 0) {
+            // The point is not "in front of" the segment. It is "away" on the p1 side
+            nearestPoint = Position.at(p1.getX(), p1.getY());
+        } else if (param > 1) {
+            // The point is not "in front of" the segment. It is "away" on the p1 side
+            nearestPoint = Position.at(p2.getX(), p2.getY());
+        } else {
+            // The point is "in front of" the segment
+            // get the projection of p on the segment
+            nearestPoint = Position.at(p1.getX() + param * p2p1Vectordx, p1.getY() + param * p2p1Vectordy);
+        }
+
+        return nearestPoint;
+    }
+
+    /**
+     * Gets the closest side of the given rectangle from the given position.
+     *
+     * @param position
+     *            the position of the point given in the rectangle upper right corner coordinates system.
+     */
+    public static Map.Entry<Position, RectangleSide> snapBorderNodeOnRectangle(Bounds borderNodeRectangle, Size parentRectangleSize) {
```

Comment:
Why is the return type a `Map.Entry`? Unless I have not understood the code of the frontend, it seems that the frontend uses a dedicated type named `SnapToRectangleInfo`. I would find it quite odd to have a more meaningful typing on the frontend than on the backend.

Reply:
done